pageserver/src/tenant/timeline.rs

```diff
@@ -926,6 +931,29 @@ impl Timeline {
         self.state.subscribe()
     }
 
+    pub async fn wait_to_become_active(
+        &self,
+        _ctx: &RequestContext, /* Prepare for use by cancellation */
+    ) -> Result<(), TimelineState> {
+        let mut receiver = self.state.subscribe();
+        loop {
+            let current_state = *receiver.borrow_and_update();
+            match current_state {
+                TimelineState::Loading => {
+                    // the sender cannot be dropped because we hold a refence to 'self'
+                    receiver.changed().await.expect("state sender dropped");
+                }
+                TimelineState::Active { .. } => {
+                    return Ok(());
+                }
+                TimelineState::Broken { .. } | TimelineState::Stopping => {
+                    // There's no chance the timeline can transition back into ::Active
+                    return Err(current_state);
+                }
+            }
+        }
+    }
```

Comment:
Follow-up task (will create it and resolve this conversation then: investigate whether we can delete `connection_manager.rs wait_for_active_timeline` in favor of this one. This one is better and in the right module.)

Reply:
https://github.com/neondatabase/neon/issues/4189